### PR TITLE
extmod/modubinascii: add newline flag to b2a_base64 method

### PR DESCRIPTION
**Pull request by chrisdecker1201**
As I wanted to use `b2a_base64` for creating a Basic-Auth header for a request I found out that the base64 is always ending with a *\n* what I don't want.

This flag also exists in the "real" python (https://docs.python.org/3/library/binascii.html#binascii.b2a_base64), so I decided to give it a try an implement it into micropython.

My current workaround is to replace the *\n* after the method call, but I don't like that.

### CODE REVIEW

**Comment by chrisdecker1201** [comment]
Probably I need some help.

1. Do you really want that change?
2. How do I fix the pipeline issues? I compiled micropython and executed the test script, and it worked. But when I execute `make test` it's failing.

**Comment by dlech** [comment]
And for cross-reference, we can check the box in https://github.com/micropython/micropython/issues/2415#issue-176546428 if this is accepted.

**Comment by chrisdecker1201** [comment]
@dlech Thank you for your review :)

The build is still failing

~~~log
TypeError: b2a_base64() takes exactly 1 positional argument (2 given)
~~~

I thought this line is enough, because I could compile and execute `micropython tests/extmod/ubinascii_b2a_base64.py`. If I understand this correct, then I define here a python method with 1 or 2 arguments.

~~~c
STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_binascii_b2a_base64_obj, 1, 2, mod_binascii_b2a_base64);
~~~

**Comment by dlech** [comment]
I guess the second argument is keyword-only in CPython, so the tests need `newline=...`. And we should probably document the difference from CPython in `tests/cpydiff/`.

**Comment by dpgeorge** [comment]
> I guess the second argument is keyword-only in CPython

Yes that's right.  So MicroPython would need to support a keyword as well, otherwise it's not possible to run the same code on both.

**Comment by chrisdecker1201** [comment]
Okay. I added `newline` as keyword

Now is not possible to write something like that

~~~python
print(binascii.b2a_base64(b"", True))
print(binascii.b2a_base64(b"f", True))
print(binascii.b2a_base64(b"fo", True))
print(binascii.b2a_base64(b"foo", True))
print(binascii.b2a_base64(b"foob", True))
print(binascii.b2a_base64(b"fooba", True))
print(binascii.b2a_base64(b"foobar", True))

print(binascii.b2a_base64(b"", False))
print(binascii.b2a_base64(b"f", False))
print(binascii.b2a_base64(b"fo", False))
print(binascii.b2a_base64(b"foo", False))
print(binascii.b2a_base64(b"foob", False))
print(binascii.b2a_base64(b"fooba", False))
print(binascii.b2a_base64(b"foobar", False))
~~~

Is it possible to add?

@dlech What do you mean with document the difference from CPython in `tests/cpydiff`?

**Comment by dpgeorge** [comment]
> Now is not possible to write something like that

That (specifying the 2nd arg as positional) is not possible in CPython so doesn't need to be possible in MicroPython either.

**Comment by dlech** [comment]
> What do you mean with document the difference from CPython in `tests/cpydiff`?

Since not using the kwarg is no longer possible in MicroPython, then there is no longer a diff to document.

**Comment by chrisdecker1201** [comment]
Is there something missing?

File: docs/library/binascii.rst [hunk 1]
```diff
@@ -31,8 +31,8 @@ Functions
    Conforms to `RFC 2045 s.6.8 <https://tools.ietf.org/html/rfc2045#section-6.8>`_.
    Returns a bytes object.
 
-.. function:: b2a_base64(data)
+.. function:: b2a_base64(data, newline=True)
```
Comment (dlech) [inline docs/library/binascii.rst:34]: Adding a `*` like this indicates that `newline` is a keyword-only argument:

```suggestion
.. function:: b2a_base64(data, *, newline=True)
```

**Comment by chrisdecker1201** [comment]
Is there anything missing? Just that I did not miss something.

File: tests/extmod/ubinascii_b2a_base64.py [hunk 2]
```diff
@@ -20,3 +20,19 @@
...
+print(binascii.b2a_base64(b"fooba", newline=True))
+print(binascii.b2a_base64(b"foobar", newline=True))
+
+print(binascii.b2a_base64(b"", newline=False))
+print(binascii.b2a_base64(b"f", newline=False))
+print(binascii.b2a_base64(b"fo", newline=False))
+print(binascii.b2a_base64(b"foo", newline=False))
+print(binascii.b2a_base64(b"foob", newline=False))
+print(binascii.b2a_base64(b"fooba", newline=False))
+print(binascii.b2a_base64(b"foobar", newline=False))
```
Comment (dpgeorge) [inline tests/extmod/ubinascii_b2a_base64.py:38]: I don't think all these extra test cases are needed.  The only new thing that needs to be tested is the new keyword argument.  So one test with it set to True, and another with it set to False, would be enough IMO.

**Comment by dpgeorge** [comment]
Squashed and merged in 2e3a2785cdd5c63788cac7ec8b5d62105ecfe4aa

Thanks for the contribution!

**Pull request closed without merging**